tensorflow/contrib/opt/python/training/elastic_average_optimizer.py

```diff
@@ -241,11 +296,29 @@ def apply_gradients(self, grads_and_vars, global_step=None, name=None):
       TypeError: If `grads_and_vars` is malformed.
       ValueError: If none of the variables have gradients.
     """
+    global_old = set(n.op.name for n in variables.global_variables())
     apply_updates = self._opt.apply_gradients(grads_and_vars)
+    global_new = set(n.op.name for n in variables.global_variables())
     with ops.control_dependencies([apply_updates]):
       local_update = state_ops.assign_add(
           self._local_step, 1, name='local_step_update').op
 
+    # this is for place the variables created by optimizer to local collection
+    # e.g., AdamOptimizer will create beta as global variables
+    def _adjust_optimizer_variable_collection(opt_vars):
+      g = ops.get_default_graph()
+      idx = 0
+      for _ in range(len(g._collections[ops.GraphKeys.GLOBAL_VARIABLES])):
+        var = g._collections[ops.GraphKeys.GLOBAL_VARIABLES][idx]
```

Comment:
Do not use _collections, use get_collection_ref instead.

However mutating these collections after the fact might not have the effect you want. Perhaps it's best to tell the optimizers to create local variables instead?

Reply:
Changed to get_collection_ref, and verified with MomentumOptimizer/AdamOptimizer/RMSPropOptimizer.